sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,668 @@
+    # define 12^x = 12 (by deductions)
+    >>> C
+    [[0, 0, 1, 2],
+     [2, 5, 2, 0],
+     [3, 1, 0, 1],
+     [4, 2, 5, 6],
+     [5, 3, 7, 8],
+     [1, 4, 6, 3],
+     [8, 7, 3, 5],
+     [6, 10, 8, 4],
+     [9, 6, 4, 7],
+     [10, 8, 10, 11],
+     [7, 9, 11, 9],
+     [11, 11, 9, 10]]
+
+    """
+    # alpha is an integer representing a "coset"
+    # since scanning can be in two cases
+    # 1. for alpha=0 and w in Y (i.e generating set of H)
+    # 2. alpha in sigma (set of live cosets), w in R (relators)
+    f = alpha
+    i = 0
+    r = len(word)
+    # list of union of generators and their inverses
+    # A = coset_table.A
+    while i < r and C[f][A.index(word.subword(i, i+1))] is not None:
+        f = C[f][A.index(word.subword(i, i+1))]
+        i += 1
+    # can this be replaced with i == r ?
+    if i >= r:
+        if f != alpha:
+            # implement the "coincidence" routine on Pg 158 of Handbook.
+            coincidence(C, f, alpha, A, p)
+            return
+    b = alpha
+    j = r - 1
+    while j >= i and C[b][A.index(word.subword(j, j+1).inverse())] is not None:
+        b = C[b][A.index(word.subword(j, j+1).inverse())]
+        j -= 1
+    if j < i:
+        # we have an incorrect completed scan with coincidence f ~ b
+        # run the "coincidence" routine
+        # line: We assume that before each call of COINCIDENCE, for α ∈ [1..n],
+        # we have p[α]=α iff α ∈ Ω.
+        print('coi')
+        p.extend(list(range(len(C))))
+        coincidence(C, f, b, A, p)
+    elif j == i:
+        print('ded')
+        # deduction process
+        C[f][A.index(word.subword(i, i+1))] = b
+        C[b][A.index(word.subword(i, i+1).inverse())] = f
+    # otherwise scan is incomplete and yields no information
+
+
+# here alpha, beta represent the pair of cosets where
+# coincidence occurs
+
+def merge(k, lamda, p, q, l):
+    phi = rep(k, p)
+    chi = rep(lamda, p)
+    if phi != chi:
+        mu = min(phi, chi)
+        v = max(phi, chi)
+        p[v] = mu
+        l += 1
+        q.append(v)
+    return l
+
+
+def rep(k, p):
+    lamda = k
+    rho = p[lamda]
+    while rho != lamda:
+        lamda = rho
+        rho = p[lamda]
+    mu = k
+    rho = p[mu]
+    while rho != lamda:
+        p[mu] = lamda
+        mu = rho
+        rho = p[mu]
+    return lamda
+
+
+# alpha, beta coincide
+def coincidence(C, alpha, beta, A, p):
+    """
+    >>> F, x, y = free_group("x, y")
+    # when coincidence occurs with 5 ~ 0, then found 4 ~ 3.
+    >>> p = [0, 1, 2, 3, 4, 5]
+    >>> C = [[1, None, None, 3], [2, 0, 2, 5],
+             [4, 1, 3, 1],[None, None, 0, 2],
+             [5, 2, 5, None], [None, 4, 1, 4]]
+    >>> coincidence(C, 0, 5, [x, x**-1, y, y**-1], p)
+    >>> C
+    [[1, 3, 1, 3],
+     [2, 0, 2, 0],
+     [3, 1, 3, 1],
+     [0, 2, 0, 2],
+     [0, 2, None, None],
+     [None, 4, 1, 4]]
+
+    """
+    l = 0
+    q = []
+    l = merge(alpha, beta, p, q, l)
+    i = 0
+    while i < l:
```

Comment:
Perhaps we could think of  `q`  as a queue (as the choice of letter seems to hint). This could then be `while len(q) > 0:`, and the next line could be `gamma = q.pop(0)`. `i` and `l` would not be needed.